pangeo_forge_orchestrator/metadata.py

```diff
@@ -0,0 +1,110 @@
+import json
+import os
+from dataclasses import dataclass, field
+from typing import Optional
+
+import fsspec
+import s3fs
+import yaml
+
+
+@dataclass
+class BakeryMetadata:
```

Comment:
Perhaps pydantic would make a better base than dataclasses.

The reason is that we have to populate our models with data from external sources, e.g. from the yaml file. Pydantic provides much more powerful validation of inputs than dataclasses.

Reply:
@rabernat As a note, we do currently do structural validation on our yaml->dict->dataclass via `dacite` in `pangeo-forge-prefect` but this is  rudimentary and does not support advanced property validation like `pydantic`.

@cisaacstern If you are going to undertake building out types with `pydantic` I would suggest using the `pangeo-forge-prefect` types as a basis and creating the types in a standalone repo we can use across modules.  I'd definitely be game for refactoring `pangeo-forge-prefect` to use them as I do think we will soon hit cases where more fine grained type validation is necessary. 👍

Reply:
> I would suggest using the pangeo-forge-prefect types as a basis 

Working on this right now and 💯 behind this suggestion: those types are _super_ helpful basis!

>  creating the types in a standalone repo

I definitely agree we want them in place where they can be reused, but I wonder if we'd be better off consolidating a lot of our orchestration tooling (i.e. anything that would need these types) into a _single_ repo of which these types would be a part, rather than continuing down the distributed repositories path we are currently on.

The ADR proposal in https://github.com/pangeo-forge/roadmap/pull/37 is basically proposing this idea, and this Minimal CLI PR is essentially a way to work through the specifics of what I want to propose in that ADR. Briefly, I see two main issues which this would address:

1. Greater intelligibility of the architecture. The wide range of packages/actions/etc. called in the CI workflow (https://raw.githubusercontent.com/pangeo-forge/flow-charts/main/renders/ci-flow-with-callstack.png) makes it hard to understand where various parts of the system live, and how therefore how to maintain/extend them.
2. Migrating CI logic out of GitHub Actions and into Python. We would still use Actions, of course, but they would just invoke logic from the central orchestration Python package, not implement that logic themselves.

This is no way a criticism of what's been built so far, but rather a reflection about adapt and grow from here.